### ISSUE TITLE
Standard format for commands & command-calling system

### ISSUE THREAD

**Issue by romgrk** [lloeki/ex-mode#49]
We should have a standard format for declaring commands, instead of throwing everything in `lib/ex.coffee`.

First draft:
```coffee
'edit':
  trigger: ['edit', 'e']
  callback: BufferCommands.edit # we could group similar commands in various classes
```

I think it could also be interesting to have regex triggers, for commands like `:.,$dx`, aka *delete everything from current line to EOF, store in register `x`*.

I'd also like to point the fact that we can use atom command-system to accomplish tasks
```coffee
dispatch = (command) ->
    atom.commands.dispatch(
        document.querySelector('atom-workspace'), command)

'enew':
  trigger: 'enew'
  callback: dispatch.bind null, 'application:new-file'
```

My last concern is argument passing. Currently, callbacks are called with `[range]` and `"args"`.
However, we might want to pass more complex objects someday, eg. `exState`, `vimState`, `editor`.
I propose to replace those arguments with a single `event` object. It could have as default the properties `.range` and `.args`. Furthermore, it would also be compatible with other event-driven systems. (eg atom commands)

**Comment by jazzpi** [comment]
I agree that we need a new format, but I'm not sure about the `trigger` being a list since `ex-mode` already match everything that is a substring of a command to that command. If it is ambiguous, `ex-mode` takes the command that comes first in alphabetical order, which isn't always the right thing, so there are some overwrites (i.e. `:tabn` being mapped to `:tabnext`, not `:tabnew`).
How does `vim` handle this?

What do you mean by `:.,$dx`? This doesn't work in `vim`. `:.,$d x` deletes into register `x`, but we don't need a new command calling system for that, `:delete` would just need to support deleting into registers. Can you elaborate on this?

**Comment by romgrk** [comment]
From vim doc:
```
:e[dit] [++opt] [+cmd]	Edit the current file.  This is useful to re-edit the
```

Short forms of commands are hard-coded, not matched from a substring search.
It needs to be defined somewhere, because trying to match `e` as a substring of something else wouldn't return `edit` as the first match. As a matter of fact, it is exactly what is done in `lib/ex.coffee`. `Ex::e` is mapped to `Ex::edit`.

For the trigger I propose the following types: `String, Regex, Array`
`String`s go in obviously. `Array`s go in for the above mentionned reason. `Regex`s go in because if we want to extend or let people extend this, it will be there if needed.

And also because some commands doesn't match the `:[range]command [args]` pattern. The best example would be the `substitute` command, which takes the form 
`:[range]s/pattern/replacement/flags`

**Comment by jazzpi** [comment]
But then, trigger would have to look like `['e', 'ed', 'edi', 'edit']`. It's certainly a possibility, but for `substitute` it would already be `['s', 'su', 'sub', 'subs', 'subst', 'substi', 'substit', 'substitu', 'substitut', 'substitute']` which is a little ugly.

From the [`ex` specification](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/ex.html#tag_20_40_13_03):
> 8. The command name shall be matched against the possible command names, and a command name that contains a prefix matching the characters specified by the user shall be the executed command. In the case of commands where the characters specified by the user could be ambiguous, the executed command shall be as follows: [...]

TL;DR `ex` does substring matching, in case of amiguity there are hardcoded exceptions. The problem with this system is however that `vim` has a lot more commands than `ex` and thus a lot more ambiguities, so I'm not sure if `vim` handles this the same way or if commands have a kind of 'priority' associated with them so that in case of ambiguity, the command with the higher priority is executed.

`substitute` does match the `:[range]command[^a-zA-Z][args]` pattern, which is basically how all commands are currently being parsed (and almost how the `ex` spec defines it). `ex` actually has an exception for `s`, `s/pattern/replacement/flags` is parsed as `s /pattern/replacement/flags`. In what situation would you need `RegEx`s to match a command?

**Comment by jazzpi** [comment]
I just noticed we would actually need to somehow pass the `vim-mode` state (or `ex-state`, since that contains a reference to `vimState`) to implement deleting to register. Is it that what you meant?

**Comment by romgrk** [comment]
Mmm, ya, you might be right about the `regex` thing not being useful.

Still, the `Array` is useful. 
Eg, for `edit` we need to hard-code both `e` and `edit`. Thus, the registration-function would parse `trigger`, and map both of them to the `editCallback`. 
So when the user inputs:
`'e'` -> call edit
`'ed'` -> do substring matching
`'edit'` -> call edit

Otherwise, we would be doing the following for each short-form of commands:
```coffee
  e: (args...) => @edit(args...)     # lib/ex.coffee: 110
```
---

> I just noticed we would actually need to somehow pass the vim-mode state (or ex-state, since that contains a reference to vimState) to implement deleting to register. Is it that what you meant?

Yes, exactly. We would be better passing an `object`, which could be
```coffee
{
  range: [int, int]
  args: String    # parsed args
  input: String   # raw input string
  vimState: VimState
  # ...
}
```
It also has the advantage that we can add more passed-values without extending the arguments list. Eg
```coffee
callback(range, args, vimState, exState, ...)
# would become
callback(event) # in all cases
```

**Comment by jazzpi** [comment]
With the `Array` for `trigger`, there could however still be unexpected results if a user adds his own command (Imagine someone adding a `:edamcheese` command, `:ed` wouldn't call `:edit` anymore).
That was my idea with 'priorities', so the command would look like:

```coffeescript
'edit':
  'priority': 1
  'callback': BufferCommands.edit
```

And if you then want to define your own commands, you choose a higher priority value to make sure you don't get in the way of pre-defined commands:

```coffeescript
'edamcheese':
  'priority': 1001
  'callback': -> console.log('The best cheese')
```

`ex-mode` then chooses the command with the lowest priority value of all matching commands.

**Comment by romgrk** [comment]
You will still need an entry
```coffee
'e':
  priority: 1
  callback: BufferCommands.edit
# in addition of
'edit':
  priority: 1
  callback: BufferCommands.edit
```
instead you could write:
```coffee
'edit':
  priority: 1
  trigger: ['edit', 'e']
  callback: BufferCommands.edit
```
then
```coffee
cmd = commands['edit'] # commands is just a descriptive object

registerCommand cmd.trigger[0], cmd
registerCommand cmd.trigger[1], cmd
registerCommand cmd.trigger[n], cmd
# where 
registerCommand = (trigger, commandDescriptionObjectOrCallbackOrWhatever) ->
  ExState.commands[trigger] = commandDescriptionObjectOrCallbackOrWhatever
```
My point is just to keep it DRY because there are a lot of shorthands.

**Comment by jazzpi** [comment]
Why do you need the entry for `'e'` then, though? Say someone types in `:e`. `ex-mode` finds all matching commands (`edit` and `edamcheese`). It then selects the one with the lowest priority value (`edit`) and executes the command linked to that. I don't really like hardcoding this stuff since it could be a lot to hardcode once a lot of commands are added.

**Comment by romgrk** [comment]
Because there are other low-priority entries. 
`'e'` is defined as `'edit'`. If it wasn't, it would first match, eg, `'echo'`.
Using a priority-system *instead* of hard-coding shorthands would be longer. And you have no reference as to which entry has more priority.

Vim doc entry, format is clear:
```help
							*:e* *:edit* *reload*
:e[dit] [++opt] [+cmd]	Edit the current file.  This is useful to re-edit the
			current file, when it has been changed outside of Vim.
			This fails when changes have been made to the current
			buffer and 'autowriteall' isn't set or the file can't
			be written.
			Also see |++opt| and |+cmd|.
			{Vi: no ++opt}
...
							*:ec* *:echo*
:ec[ho] {expr1} ..	Echoes each {expr1}, with a space in between.  The
			first {expr1} starts on a new line.
			Also see |:comment|.
			Use "\n" to start a new line.  Use "\r" to move the
			cursor to the first column.
			Uses the highlighting set by the |:echohl| command.
			Cannot be followed by a comment.

```

**Comment by lloeki** [comment]
Well, there's so much going on here. I'm following this and have a number of ideas, but let me ponder this a bit more. The night's your ally :-)
